### DISCUSSION TITLE
Correct way to use DirectedGraph layout for subset of elements

### DISCUSSION THREAD

**Discussion by Sagariouz** [clientIO/joint#2268] [Q&A]
### Introduction

We are using Jointjs to create an uml diagram functionality. It is important for the user to move elements around to their liking and we persist the position(s) as the user works. When an user adds a new element, we position it using mouse's x and y coordinates.

But we want to make use of DirectedGraph layout when a bunch of elements/links needs to be added, but not to change  the positions of the current placed elements. Got it to work to an extent, but elements start to overlap.

Any other suggestions, or I may not understand DirectedGraph layout fully?
To summate, I would like to apply the layout only to items that don't a set position yet

### Steps to reproduce

What I've tried:
- Call joint.layout.DirectedGraph.layout() with input of only the newly added elements and their links, I realised that I needed to add the links as well, otherwise the layout strategy didn't work - overlap occurs.
- Call joint.layout.DirectedGraph.layout() with input of the full graph, but override the setPosition() to only set the position of the newly added elements, and ignore the allready placed elements - overlap occurs.
- Calling just joint.layout.DirectedGraph.layout() with the full graph, may move elements that the user placed at a certain point, which we don't want.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment]
> Call joint.layout.DirectedGraph.layout() with input of only the newly added elements and their links, I realised that I needed to add the links as well, otherwise the layout strategy didn't work - overlap occurs.

Do you mean that the newly added elements are overlapping with the existing ones after the layout?

**Reply by Sagariouz** [reply]
Correct, or that was my main premise, but if I look at the below, these are all new - so my premise might be wrong.
![image](https://github.com/clientIO/joint/assets/1581542/659efb96-308a-4756-a0aa-334143fb03f7)

There might be something else at play, I'll investigate further.
It would be great to have a firm recommended direction of how I would be able to accomplish this, then I'll focus on that approach?

**Reply by kumilingus** [reply]
It's not possible to set restrictions on element positions with [Dagre](https://github.com/dagrejs/dagre/wiki) (Directed Graph Layout).

This StackOverflow answer is true for Dagre: https://stackoverflow.com/a/50217292/2921495

_Note: Before adding new elements you can get the bounding box of the graph and then place the new elements next to, above or below (depending on the `rankDir` used) the existing graph so that the elements do not overlap._
